### ISSUE TITLE
[Nagara][PDX224] Endless adsprpcd initialisation failure loop  on Android 13 

### ISSUE THREAD

**Issue by faenil** [sonyxperiadev/bug_tracker#795]
**Platform:** Nagara
**Device:** pdx224 (Xperia 5 IV)
**Kernel version:** 5.10 from `android-13.0.0_r30` branch ([f3f309be38e952937ce363d5e7201934c56ced88](https://github.com/sonyxperiadev/kernel/commit/f3f309be38e952937ce363d5e7201934c56ced88))
**Android version:** `android-13.0.0_r30`
**Software binaries version:**  `64.0.H.4.18` + `SW_binaries_for_Xperia_Android_13_5.10_v1b_nagara.zip`

Source code used:

- Kernel source is untouched, just rebuilt via `build-clang-shared.sh` (prebuilt kernel for this device is not yet released)
- Nagara device repo is basically `t-mr1` branch plus https://github.com/sonyxperiadev/device-sony-nagara/compare/t-mr1...MarijnS95:device-sony-nagara:t-mr1
- AOSP branch `android-13.0.0_r30`, repo_update'd and a2dp patches applied (see bullet list below)
  - undid https://github.com/sonyxperiadev/repo_update/commit/3f90fdcb2f8c723e3e8f981c976d7e8424719969 
  - applied the [device-sony-common](https://github.com/sonyxperiadev/device-sony-common/pull/899) build fix
  - applied the [a2dp device PR](https://github.com/sonyxperiadev/device-sony-nagara/commit/55f46299561c389e64f0ff55ce25dc01e84ce2af)

The target is `-eng` variant

**Previously working on**
n/a

**Description**
Logs show adsprpcd keeps spinning in an endless initialisation loop that never succeeds and which runs every 25msecs.

**Symptoms**
The device overheats and has to be turned off after 5-10minutes.
The device is hence unusable and is too hot to be held in your hand.

**How to reproduce**
Build and flash the image using the source code shared above.
Boot the device.
Wait for a few minutes and notice the temperature skyrocket.

**Additional context**

Note: these logcat/dmesg logs contain additional debug logs that I added in the `adsprpc.c` driver
[adsprpcd_init_loop.zip](https://github.com/sonyxperiadev/bug_tracker/files/11142897/adsprpcd_init_loop.zip)

**Comment by faenil** [comment]
I investigated this on and off over the last few weekends:

started with `strace`'ing `adsprcpd`, which highlighted the failing syscall
```
writev(3, [{iov_base="\0\213\3\237\3\370cZyn\31", iov_len=11}, {iov_base="\3", iov_len=1}, {iov_base="adsprpcd\0", iov_len=9}, {iov_base="vendor/qcom/proprietary/adsprpc/"..., iov_len=85}], 4) = 106
getuid()                                = 1013
writev(4, [{iov_base="lq\0\365\3\213\3", iov_len=7}, {iov_base="\0\213\3\237\3\370cZyn\31", iov_len=11}, {iov_base="\3", iov_len=1}, {iov_base="adsprpcd\0", iov_len=9}, {iov_base="vendor/qcom/proprietary/adsprpc/"..., iov_len=85}], 5) = 113
getuid()                                = 1013
writev(3, [{iov_base="\0\213\3\237\3\370cj\304u\31", iov_len=11}, {iov_base="\6", iov_len=1}, {iov_base="adsprpcd\0", iov_len=9}, {iov_base="vendor/qcom/proprietary/adsprpc/"..., iov_len=100}], 4) = 121
getuid()                                = 1013
writev(4, [{iov_base="l\200\0\365\3\213\3", iov_len=7}, {iov_base="\0\213\3\237\3\370cj\304u\31", iov_len=11}, {iov_base="\6", iov_len=1}, {iov_base="adsprpcd\0", iov_len=9}, {iov_base="vendor/qcom/proprietary/adsprpc/"..., iov_len=100}], 5) = 128
getuid()                                = 1013
writev(3, [{iov_base="\0\213\3\237\3\370c\2562}\31", iov_len=11}, {iov_base="\4", iov_len=1}, {iov_base="adsprpcd\0", iov_len=9}, {iov_base="vendor/qcom/proprietary/adsprpc/"..., iov_len=166}], 4) = 187
getuid()                                = 1013
writev(4, [{iov_base="l\302\0\365\3\213\3", iov_len=7}, {iov_base="\0\213\3\237\3\370c\2562}\31", iov_len=11}, {iov_base="\4", iov_len=1}, {iov_base="adsprpcd\0", iov_len=9}, {iov_base="vendor/qcom/proprietary/adsprpc/"..., iov_len=166}], 5) = 194
futex(0x6fb92f8fc0, FUTEX_WAKE_PRIVATE, 2147483647) = 0
openat(AT_FDCWD, "/dev/adsprpc-smd-secure", O_RDONLY|O_NONBLOCK) = 7
ioctl(7, _IOC(_IOC_READ|_IOC_WRITE, 0x52, 0x8, 0x4), 0x7ff2c22cd8) = 0
ioctl(6, _IOC(_IOC_READ|_IOC_WRITE, 0x48, 0, 0x18), 0x7ff2c22ad0) = 0
mmap(NULL, 4096, PROT_READ|PROT_WRITE, MAP_SHARED, 8, 0) = 0x724e737000
ioctl(7, _IOC(_IOC_READ|_IOC_WRITE, 0x52, 0x6, 0x28), 0x7ff2c22ce0) = -1 ENODEV (No such device)
munmap(0x724e737000, 4096)              = 0
```

the `ioctl` returning `ENODEV` can be found in the kernel source as:
`#define FASTRPC_IOCTL_INIT      _IOWR('R', 6, struct fastrpc_ioctl_init)`

which confirms this is related to the init procedure.

That ioctl number macro is equal (after unpacking all macros) to ioctl number `3223867910`

Adding further debug logs revealed that the failure comes from `fastrpc_channel_open` (note `-19` is `-ENODEV`)
```
02-16 12:59:25.008  1007  1007 I adsprpcd: vendor/qcom/proprietary/adsprpc/src/fastrpc_apps_user.c:2598: domain_deinit for domain 4: dev -1
02-16 12:59:25.008  1007  1007 I adsprpcd: vendor/qcom/proprietary/adsprpc/src/fastrpc_apps_user.c:2598: domain_deinit for domain 5: dev -1
02-16 12:59:25.008  1007  1007 I adsprpcd: vendor/qcom/proprietary/adsprpc/src/fastrpc_apps_user.c:2598: domain_deinit for domain 6: dev -1
02-16 12:59:25.008  1007  1007 I adsprpcd: vendor/qcom/proprietary/adsprpc/src/fastrpc_apps_user.c:2598: domain_deinit for domain 7: dev -1
02-16 12:59:25.008  1007  1007 I adsprpcd: vendor/qcom/proprietary/adsprpc/src/rpcmem_android.c:322: rpcmem_deinit_internal done
02-16 12:59:25.008  1007  1007 I adsprpcd: vendor/qcom/proprietary/adsprpc/src/fastrpc_apps_user.c:3349: fastrpc_apps_user_deinit done
02-16 12:59:25.008  1007  1007 E adsprpcd: vendor/qcom/proprietary/adsprpc/src/adsprpcd.c:53:adsp daemon will restart after 25ms...
02-16 12:59:25.007  1007  1007 I         : ioctl ioctl num 3221508616 549053674728 mmap ioctl -1071623678
02-16 12:59:25.014  1007  1007 I         : fastrpc_check_pd_status bailing out: err 0
02-16 12:59:25.014  1007  1007 I         : ioctl ioctl num 3223867910 549053674736 mmap ioctl -1071623678
02-16 12:59:25.014  1007  1007 I         : fastrpc_check_pd_status bailing out: err 0
02-16 12:59:25.014  1007  1007 I         : fastrpc_init_process start
02-16 12:59:25.014  1007  1007 I         : fastrpc_channel_open start
02-16 12:59:25.014  1007  1007 I         : fastrpc_channel_open failed lpass
02-16 12:59:25.014  1007  1007 I         : fastrpc_init_process channel open failed
02-16 12:59:25.014  1007  1007 I fastrpc_init_process: bailing out with err -19
```

which is failing due to a null `rpdev` for the channel subsystem ([see the failing point here](https://github.com/sonyxperiadev/kernel/blob/f3f309be38e952937ce363d5e7201934c56ced88/drivers/char/adsprpc.c#L5880))

That is symptom that the subsystem might have failed to initialised.

Indeed, logs also show
```
02-16 13:01:16.937     0     0 I         : Booting Linux on physical CPU 0x0000000000 [0x410fd462]
02-16 13:01:16.937     0     0 I         : Linux version 5.10.66-sodp-gf3f309be38e9-dirty (faenil@faenil-PB50-70RF-RD-RC) (Android (8490178, based on r450784d) clang version 14.0.6 (https://android.googlesource.com/toolchain/llvm-project 4c603efb0cca074e9238af8b4106c30add4418f6), LLD 14.0.6) #3 SMP PREEMPT Sat Mar 25 14:58:56 GMT 2023
...
...
02-16 13:01:23.392    78    78 I adsp-loader soc: qcom,msm-adsp-loader: adsp_load_fw: Load default ADSP
02-16 13:01:23.392    78    78 I remoteproc1: powering up 3000000.remoteproc-adsp
...
02-16 13:01:23.413   879   879 I ueventd : firmware: loading 'adsp.mdt' for '/devices/platform/soc/3000000.remoteproc-adsp/remoteproc/remoteproc1/adsp.mdt'
02-16 13:01:23.416     1     1 I init    : starting service 'incidentd'...
02-16 13:01:23.416   879   879 I ueventd : loading /devices/platform/soc/3000000.remoteproc-adsp/remoteproc/remoteproc1/adsp.mdt took 2ms
02-16 13:01:23.416    78    78 I remoteproc1: Booting fw image adsp.mdt, size 5324
02-16 13:01:23.420    78    78 W qcom_q6v5_pas 3000000.remoteproc-adsp: Direct firmware load for adsp.b24 failed with error -2
02-16 13:01:23.420     1     1 I init    : starting service 'installd'...
02-16 13:01:23.430    78    78 W qcom_q6v5_pas 3000000.remoteproc-adsp: Falling back to sysfs fallback for: adsp.b24
02-16 13:01:23.433   883   883 I ueventd : firmware: loading 'adsp.b24' for '/devices/platform/soc/3000000.remoteproc-adsp/firmware/adsp.b24'
02-16 13:01:23.442   883   883 I ueventd : loading /devices/platform/soc/3000000.remoteproc-adsp/firmware/adsp.b24 took 8ms
...
02-16 13:01:23.491   911   911 I ueventd : loading /devices/platform/soc/3000000.remoteproc-adsp/firmware/adsp.b01 took 0ms
...
...
02-16 13:01:23.869  1047  1047 I ueventd : loading /devices/platform/soc/3000000.remoteproc-adsp/firmware/adsp.b11 took 3ms
...
02-16 13:01:23.901  1056  1056 I ueventd : loading /devices/platform/soc/3000000.remoteproc-adsp/firmware/adsp.b12 took 4ms
...
...
LOADING LAST FW FILE...
...
02-16 13:01:24.165  1109  1109 I ueventd : loading /devices/platform/soc/3000000.remoteproc-adsp/firmware/adsp.b22 took 5ms
02-16 13:01:24.166    78    78 E 3000000.remoteproc-adsp: glink-edge: failed to register early notif -524
...
02-16 13:01:24.176    78    78 E qcom_q6v5_pas 3000000.remoteproc-adsp: Bus scaling not setup for 3000000.remoteproc-adsp
...
02-16 13:01:24.310     0     0 W         : [    C0] IRQ 407 device glink-native-adsp returned IRQ_WAKE_THREAD but no thread function available.
...
02-16 13:01:24.321   179   179 E qcom_q6v5_pas 3000000.remoteproc-adsp: Bus scaling not setup for 3000000.remoteproc-adsp
..
02-16 13:01:24.333    78    78 E genirq  : Flags mismatch irq 407. 00006001 (glink-native-adsp) vs. 00004081 (glink-native-adsp)
02-16 13:01:24.333    73    73 I qcom_smd_qrtr_probe: SMD QRTR driver probed
02-16 13:01:24.343    78    78 E 3000000.remoteproc-adsp: glink-edge: failed to request IRQ with -16
02-16 13:01:24.343    78    78 E remoteproc1: failed to probe subdevices for 3000000.remoteproc-adsp: -16
...
...
02-16 13:01:24.402    78    78 I adsprpc : fastrpc_restart_notifier_cb: received RAMDUMP notification for lpass
02-16 13:01:24.402    78    78 E adsp-loader soc: qcom,msm-adsp-loader: adsp_load_fw: pil get failed,
02-16 13:01:24.410    78    78 E adsp-loader soc: qcom,msm-adsp-loader: adsp_load_fw: Q6 image loading failed
```

This last log shows the root cause of the failure, i.e. **there is a failure in requesting the IRQ**!
```
02-16 13:01:24.333    78    78 E genirq  : Flags mismatch irq 407. 00006001 (glink-native-adsp) vs. 00004081 (glink-native-adsp)
...
02-16 13:01:24.343    78    78 E 3000000.remoteproc-adsp: glink-edge: failed to request IRQ with -16
02-16 13:01:24.343    78    78 E remoteproc1: failed to probe subdevices for 3000000.remoteproc-adsp: -16
```

**Comment by faenil** [comment]
After instrumenting the IRQ request code, it turned out `glink-native` is, indeed, requesting the same IRQ with different flags.

[Here, in qcom_glink_native_start](https://github.com/sonyxperiadev/kernel/blame/f3f309be38e952937ce363d5e7201934c56ced88/drivers/rpmsg/qcom_glink_native.c#L2283) but also in [qcom_glink_native_probe](https://github.com/sonyxperiadev/kernel/blame/f3f309be38e952937ce363d5e7201934c56ced88/drivers/rpmsg/qcom_glink_native.c#L2226), where different flags are used.

**Comment by faenil** [comment]
I have verified that reverting https://github.com/sonyxperiadev/kernel/commit/d2fabd8d03f2679dfca23a7f433df9189951ab27 fixes the `adsprpcd` init failure loop :rocket:

**Comment by jerpelea** [comment]
please ifdef the change so that it is excluded for nagara

Thanks 
Alin

**Comment by MarijnS95** [comment]
Thanks for finding all this! Indeed, looks like @konradybcio needs to give a bit more clarification on https://github.com/sonyxperiadev/kernel/commit/d2fabd8d03f2679dfca23a7f433df9189951ab27, which is just a random amalgamation of the code around the original `devm_request_irq` prior to https://github.com/sonyxperiadev/kernel/commit/b21bd352af382264b3d5b7acb2be24e7794a8f17 and https://github.com/sonyxperiadev/kernel/commit/f551bd17e4fe7df365799ebaecf2131876537402.

Let's revert it for now (with the above as context), and we'll redo the port in a sane way when (not if) it turns out to break 6375.

**Comment by faenil** [comment]
Great, thanks @MarijnS95, that's exactly what I thought, great to see we're on the same page

**Comment by MarijnS95** [comment]
Looks like GitHub did not auto-close this (nor link the PR and issue together) because the `Fixes:` link is a labeled rather than plain URL: `Fixes: [#795](https://github.com/sonyxperiadev/bug_tracker/issues/795)`. This also confused me for a sec as a plain `#795` would reference that number _within the `kernel` repository_, otherwise GitHub renders it as `sonyxperiadev/bug_tracker#795` automatically.

**Comment by faenil** [comment]
>This also confused me for a sec as a plain #795 would reference that number within the kernel repository

Yep, that confused me as well.

I see that `sonyxperiadev/bug_tracker#795` [is the way to go](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) :+1:
